controllers/virtualmachine/storagepolicyusage/storagepolicyusage_controller.go

```diff
@@ -272,3 +368,64 @@ func reportReserved(
 
 	return nil
 }
+
+// reportReservedForSnapshot reports the reserved capacity for a snapshot.
+func (r *Reconciler) reportReservedForSnapshot(
+	vmSnapshot vmopv1.VirtualMachineSnapshot,
+	total *resource.Quantity,
+	storageClass string) error {
+
+	if vmSnapshot.Spec.VMRef == nil {
+		return fmt.Errorf("vmRef is not set")
+	}
+
+	// It's possible that the snapshot's requested capacity is not updated yet,
+	// so we skip the calculation for now
+	if vmSnapshot.Status.Requested == nil {
+		r.Logger.V(5).Info("snapshot has no requested capacity reported by controller yet, skip calculating reserved capacity for it", "snapshot", vmSnapshot.Name)
+		return nil
+	}
+	// Loop through all the requested capacities and add the requested capacity
+	// with the same storage class as the SPU's storage class.
+	for _, requested := range vmSnapshot.Status.Requested {
+		if requested.StorageClass == storageClass {
+			r.Logger.V(5).Info("adding snapshot's requested capacity to total reserved capacity", "snapshot", vmSnapshot.Name, "requested", requested.Total)
+			total.Add(*requested.Total)
+			break
+		}
+	}
+
+	return nil
+}
+
+// reportUsedForSnapshot reports the used capacity for a snapshot.
+// Get the snapshot size from the VMSnapshot's status.Used.Total.
+func (r *Reconciler) reportUsedForSnapshot(
+	vmSnapshot vmopv1.VirtualMachineSnapshot,
+	total *resource.Quantity,
+	storageClass string) error {
+
+	if vmSnapshot.Spec.VMRef == nil {
+		return fmt.Errorf("vmRef is not set")
+	}
+
+	vm := &vmopv1.VirtualMachine{}
+	if err := r.Client.Get(r.Context, client.ObjectKey{Namespace: vmSnapshot.Namespace, Name: vmSnapshot.Spec.VMRef.Name}, vm); err != nil {
```

Comment:
We've already done a List of the VMs. Can we refactor this to use that. Also, this going repeating fetch the same VM if it has multiple snapshots

The funky and unfortunate decision to try to push quota managed in k8s means that this reconcile is called all the time so it needs to be very mindful

Reply:
>We've already done a List of the VMs. Can we refactor this to use that. Also, this going repeating fetch the same VM if it has multiple snapshots

Yep it's doable. There is still the case that VM has different storageclass but some of its PVC has the target storageclass, then in that case we still need to fetch the VM.

>The funky and unfortunate decision to try to push quota managed in k8s means that this reconcile is called all the time so it needs to be very mindful

Just curious, how would you do it differently. rely on the external system manager quota and reports failure of creating resource then just reflect that in CR, so that nothing will need to query all resource in a single namespace at a single time?

Reply:
>I'm confused if we've done a list of the VMs why is a get of the VM needed for PVCs?
Oh wait, PVC's disk usage is only considered when calculating SPU Reservation, not in SPU Usage. Ok then nvm.